src/delivery_problems/mod.rs

```diff
@@ -101,8 +105,35 @@ where
         bounce_type: BounceType,
         bounce_subtype: BounceSubtype,
     ) -> AppResult<()> {
+        let problem_type: ProblemType = From::from(bounce_type);
+        let problem_subtype: ProblemSubtype = From::from(bounce_subtype);
         self.auth_db
-            .create_bounce(address, From::from(bounce_type), From::from(bounce_subtype))?;
+            .create_bounce(address, problem_type, problem_subtype)?;
+        self.record_delivery_problem(address, problem_type, problem_subtype)?;
+        Ok(())
+    }
+
+    fn record_delivery_problem(
+        &self,
+        address: &EmailAddress,
+        problem_type: ProblemType,
+        problem_subtype: ProblemSubtype,
+    ) -> AppResult<()> {
+        let mut problems: Vec<DeliveryProblem> = self
+            .db
+            .get(address.as_ref(), DataType::DeliveryProblem)?
+            .unwrap_or_else(|| Vec::new());
+        problems.insert(
+            0,
```

Comment:
IIUC, inserting at the head of a `Vec` like this going to be O(N) in the number of delivery problems in the list, because it has to shift every other element to the right.  Is it possible to append to the end and then traverse the vector in reverse order, or something like that?

What will happen if we have so many bounce records for an address that the serialized value exceeds the maximum size of a string in redis?  Should we purge older entries to keep the size of the list under control?

We expect the size of the list to be small in practice, but we don't want to open any DoS vectors here.

(I know it would be a much bigger PR, but part of me wonders whether we should use redis's builtin list datatype for this...)

Reply:
My plan was to purge older items, as per the discussion at https://github.com/mozilla/fxa-email-service/issues/166#issuecomment-425035671. But I can look at appending and then changing the logic that processes the list instead, no probs.

Althoug also, just fwiw, it wouldn't be possible to DoS the service from this code as it is not invoked from the API. We'd DoS the queues process instead, meaning we'd be slow to process bounce/complaint/delivery notification.

Reply:
Actually, is it true that we want the items in this list to be sorted by timestamp?  Inserting at the head or tail of the list won't necessarily guarantee that, if the SQS processing happens to deal with messages out of order (which it can).  Perhaps we should explicitly insert in sorted order here to be sure.

Reply:
Although I guess we just generate the timestamp ourselves don't we, rather than taking a timestamp from the message....

I should probably stop worrying at this point and just leave it in your capable hands :-)

Reply:
Yeah, timestamps are generated in the auth db at the moment, presumably because we decided it was risky to trust the time reported by SES. I'm in favour of revisiting that decision as part of this broader migration process fwiw, we're already trusting SES to report accurate bounce/complaint events to us so trusting the timestamp too doesn't seem too big a leap.

Reply:
> I'm in favour of revisiting that...

Ref: #205